cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,230 @@
+# Design Note: Environments
+
+## Metadata
+
+- Author: @ellismg
+- Date: 2022-07-27
+
+## Overview
+
+This document outlines what an environment is in `azd` today and outlines a change in model to remove the need to introduce the concept on the critical path for a user getting started with `azd` to manage their application.
+
+## Current Design
+
+The concept on an environment has been present in `azd` for a long while, but the way it has been used has changed over time. In addition, the ties between an environment name and provisioned resources has changed over time.
+
+Every command in `azd` today runs in the context of an environment. The name of the environment can be provided by the `-e` flag which is command to all `azd` commands and when not set, the *default environment* is used. The name of the currently selected environment is stored in `.azure/config.json` file, under the `defaultEnvironment`. The *default environment* can set with `azd env select [environment-name]`. If no environment name is passed and no default environment has been selected then the user is prompted to create a new environment before the operation can proceed.
+
+During environment creation the user is prompted for the following pieces of information:
+
+1. A Name for the environment.
+2. An Azure location.
+3. An Azure subscription.
+
+Names of environments are validated against the regular expression `[a-zA-Z0-9-\(\)_\.]{1,64}$` (this matches the regular expression used by ARM to validate names of ARM deployments).
+
+The name of the environment is used to name a directory under `.azure` which contains files related to the environment. The main file in this folder is the `.env` file, which is used to store key/value pairs in the environment.
+
+The values selected during initialization are stored in this `.env` file with well known key names:
+
+- `AZURE_ENV_NAME`: The environment name.
+- `AZURE_LOCATION`: The short location name (e.g. `eastus2`)
+- `AZURE_SUBSCRIPTION_ID`: The id of the selected subscription (e.g. `faa080af-c1d8-40ad-9cce-e1a450ca5b57`)
+
+In addition, the following value is added without any user interaction:
+
+- `AZURE_PRINCIPAL_ID`: The id of the current principal authenticated to the `az` CLI.
+
+All of these values must be present in an environment for it to be considered initialized.  If any are missing (for example, by editing the .env file) the user will be prompted the next time `azd` runs to configure these values. Note that this will require the user be logged into the `az` CLI since configuring these values requires the CLI (to list locations, subscriptions and the object id of the current principal).
```

Comment:
Yes - this is true today and enforced by [`ensureEnvironmentInitialized`](https://github.com/Azure/azure-dev/blob/77ab01c3bcf40d12852d543506ecbb935a779f81/cli/azd/cmd/util.go#L168-L256).

I agree that it is not /technically/ required - you don't need this piece of information to create an azure deployment and our bicep templates are written such that if this value was unset things would deploy successfully. It would break our end to end experience where we run the app locally but connect to the cosmos db instance in the cloud (and other things like KeyVault where we expect to leverage managed identity and create access rules as part of infrastructure deployment).

I suspect this is the reason we ended up treating it as something that should always be present (the fact that the value is fixed and doesn't reflect the "current" principal but rather the principal we had when we initialized the environment is likely a bug in the current implementation we should fix if we don't adopt this design.